src/Frame/Frame.cc

```diff
@@ -2334,4 +2331,216 @@ bool Frame::GetDownsampledRasterData(
         tile_data.data(), data.data(), tile_original_width, tile_original_height, downsampled_width, downsampled_height, 0, 0, mip);
 }
 
+bool Frame::CalculateVectorField(const std::function<void(CARTA::VectorOverlayTileData&)>& callback) {
+    auto& settings = _vector_field_settings;
+    if (settings.smoothing_factor < 1) {
+        return true;
+    }
+
+    if (settings.stokes_intensity < 0 && settings.stokes_angle < 0) {
+        _vector_field_settings.ClearSettings();
+        auto empty_response = Message::VectorOverlayTileData(settings.file_id, _z_index, settings.stokes_intensity, settings.stokes_angle,
+            settings.compression_type, settings.compression_quality);
+        empty_response.set_progress(1.0);
+        callback(empty_response);
+        return true;
+    }
+
+    return DoVectorFieldCalculation(settings, callback);
+}
+
+bool Frame::DoVectorFieldCalculation(
```

Comment:
This is a very long function.  See what you can move into smaller functions, or into VectorSettings if it does not involve `GetDownsampledRasterData` (for example, the calculations starting with the lambda functions do not need to be in Frame)